examples/vision/transformer_in_transformer.py

```diff
@@ -0,0 +1,328 @@
+    # create multiple layers of the TNT block.
+    for _ in range(num_transformer_layer):
+        pixel_embedding, patch_embedding = transformer_in_transformer_block(
+            pixel_embedding,
+            patch_embedding,
+            outer_block_embedding_dim,
+            inner_block_embedding_dim,
+            num_pixel,
+            outer_block_num_heads,
+            inner_block_num_heads,
+            mlp_ratio,
+            attention_dropout,
+            projection_dropout,
+        )
+    patch_embedding = layers.LayerNormalization(epsilon=1e-5)(patch_embedding)
+    x = layers.GlobalAvgPool1D()(patch_embedding)
+    outputs = layers.Dense(num_classes, activation="softmax")(x)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    return model
+
+
+"""
+## Train on CIFAR-100
+"""
+
+model = get_model()
+model.compile(
+    loss=keras.losses.CategoricalCrossentropy(label_smoothing=label_smoothing),
+    optimizer=tfa.optimizers.AdamW(
+        learning_rate=learning_rate, weight_decay=weight_decay
+    ),
+    metrics=[
+        keras.metrics.CategoricalAccuracy(name="accuracy"),
+        keras.metrics.TopKCategoricalAccuracy(5, name="top-5-accuracy"),
+    ],
+)
+
+history = model.fit(
+    x_train,
+    y_train,
+    batch_size=batch_size,
+    epochs=num_epochs,
+    validation_split=validation_split,
+)
+
+"""
+### Visualize the training progress of the model.
+"""
+
+plt.plot(history.history["loss"], label="train_loss")
+plt.plot(history.history["val_loss"], label="val_loss")
+plt.xlabel("Epochs")
+plt.ylabel("Loss")
+plt.title("Train and Validation Losses Over Epochs", fontsize=14)
+plt.legend()
+plt.grid()
+plt.show()
+
+plt.plot(history.history["accuracy"], label="train_accuracy")
+plt.plot(history.history["val_accuracy"], label="val_accuracy")
+plt.xlabel("Epochs")
+plt.ylabel("Accuracy")
+plt.title("Train and Validation Accuracies Over Epochs", fontsize=14)
+plt.legend()
+plt.grid()
+plt.show()
+
+"""
+### Let's display the final results of the test on CIFAR-100.
+"""
+
+loss, accuracy, top_5_accuracy = model.evaluate(x_test, y_test)
+print(f"Test loss: {round(loss, 2)}")
+print(f"Test accuracy: {round(accuracy * 100, 2)}%")
+print(f"Test top 5 accuracy: {round(top_5_accuracy * 100, 2)}%")
+
+"""
+After 50 epochs, the TNT model achieves around 48% accuracy and
+78% top-5 accuracy on the test data. It only has 0.8M parameters.
+From the above loss curve, we can find that the model gradually converges,
+but it never achieves state of the art performance. We can apply some data augmentation
+to
+obtain better performance, like RandAugment, MixUp etc. We also can adjust the depth
+of model, learning rate or increase the size of embedding.Compared to the conventional
+vision transformers [(ViT)](https://arxiv.org/abs/2010.11929) which corrupts the local
+structure
+of the patch, the TNT can better preserve and model the local information
+for visual recognition.
+"""
```

Comment:
If you know of any good links for further reading to go deeper, feel free to include them at the end of the example.